### DISCUSSION TITLE
How can I attach ld file with lld or mold?

### DISCUSSION THREAD

**Discussion by bossjisu** [rui314/mold#1070] [Q&A] [answered]
A.ld
```
SECTIONS
{
    .rel.rodata.func_reg : {
        KEEP(*(INIT_REG_ENTRY_START));
        PROVIDE(init_reg0_start = .);
        *(.init_reg.early.*)
        PROVIDE(init_reg0_end = .);
        ...
    }
}
INSERT AFTER .text;
```

A.hpp
```
typedef bool (*init_func_type)();
#define ADD_INIT(idx, func) \
    static init_func_type i_##idx##func __attribute__((__section__(".init_reg." #idx "." #func))) = (func);
```

A.cpp
```
static bool init_early() {
    std::cout << "init_early" << std::endl;
    return true;
}

ADD_INIT(early, init_early) // Expecting init_early() pointer places at ".init_reg.early.init_early"
```

B.cpp
```
extern uint64_t init_reg0_start[];
extern uint64_t init_reg0_end[];
uint64_t init_size = init_reg0_end - init_reg0_start;
for (uint64_t i = 0; i < init_size; j += sizeof(init_func_type)) {
    auto init_func = *(init_func_type*)(init_reg0_start + i);
    init_func(); // Expecting execution of init_early() function in A.cpp
}
```

I have this implementation on my project using linker script, and I attach the A.ld with the CMAKE option below.

`target_link_options(${BIN_TARGET} PRIVATE -Wl,-T${CMAKE_CURRENT_SOURCE_DIR}/A.ld)`

It works well with gcc linker but it doesn't work with lld and mold.
It doesn't give me a build failure but init_size at B.cpp is always 0 and I don't see pointers at that section.
How should I fix it to make it work? Please give me an advice...
There can be some typo or missing things to make the example short, but I hope it explains my purpose well enough.
Always welcomed if you need more explanations!

Thanks for reading!

**Comment by rui314** [comment, marked as answer]
You don't need to use a linker script to do what you want to do.

Just place all initialization functions to `init_reg` section (notice the lack of the leading `.`), then the linker concatenate input object file's `init_reg` sections and create a single "init_reg" section in an output file. You can refer to the beginning and ending of the section with `__start_init_reg` and `__stop_init_reg` linker-generated symbols, respectively.

But perhaps you don't even need the above mechanisms. Are you aware of GCC's `__attribute__((constructor))`? With that, you can let the dynamic loader to call arbitrary functions before main.
